### PR TITLE
Feature/set host

### PR DESCRIPTION
**Pull request by rrahn**
Adds setHost with move semantics to replace old non-const and const version which was misused in many places, especially in the Pattern implementation If 
If c++11 is enabled! The setHost of Pattern delegates now correctly to a Holder, meaning that temporaries are copied and lvalue refs are set dependent.

### CODE REVIEW

File: include/seqan/find/find_hamming_simple.h [hunk 1]
```diff
@@ -78,11 +78,26 @@ class Pattern<TNeedle, HammingSimple> {
...
+#ifdef SEQAN_CXX11_STANDARD
+    template <typename TNeedle2>
+    Pattern(TNeedle2 && ndl,
+            int k = -1,
+            SEQAN_CTOR_DISABLE_IF(IsSameType<typename std::remove_reference<TNeedle2>::type const &, Pattern const &>))
+                :   maxDistance(-1),
+                    distance(0),
+                    matchNFlags(0)
+    {
+        setHost(*this, std::forward<TNeedle2>(ndl), k);
```
Comment (rrahn) [inline include/seqan/find/find_hamming_simple.h]: Actually setting the distance should not be done using the setHost interface.
I think there are interfaces to set the distance. We rather keep a clean setHost interface and set another feature with another function, especially as the current interface does not require the distance parameter for the initialisation of the Pattern.

File: include/seqan/find/find_multiple_bfam.h [hunk 2]
```diff
@@ -110,16 +110,29 @@ class Pattern<TNeedle, MultiBfam<TSpec> >
...
     {
         SEQAN_CHECKPOINT
         setHost(*this, ndl);
     }
+# endif
 
     ~Pattern()
-    {
-    }
+    {}
```
Comment (rrahn) [inline include/seqan/find/find_multiple_bfam.h]: Remove default destructor.

File: include/seqan/find/find_myers_ukkonen.h [hunk 3]
```diff
@@ -615,109 +674,100 @@ _patternMatchNOfFinder(Pattern<TNeedle, Myers<TSpec, THasState, void> > & patter
...
+//_myersSetHost(Pattern<TNeedle, Myers<TSpec, THasState, TFindBeginPatternSpec> > &, TNeedle2 const &)
+//{
+//}
+//
+//template <typename TNeedle, typename TSpec, typename TFinderCSP, typename TPatternCSP, typename THasState, typename TFindBeginPatternSpec, typename TNeedle2>
+//inline void
+//_myersSetHost(Pattern<TNeedle, Myers<AlignTextBanded<TSpec, TFinderCSP, TPatternCSP>, THasState, TFindBeginPatternSpec> > & pattern, TNeedle2 const & ndl)
+//{
+//    setValue(pattern.data_host, ndl);
+//}
```
Comment (rrahn) [inline include/seqan/find/find_myers_ukkonen.h]: remove commented block

File: include/seqan/align/dp_matrix.h [hunk 4]
```diff
@@ -106,29 +113,29 @@ class DPMatrix_<TValue, FullDPMatrix>
...
+//    ~DPMatrix_() {}
+//
+//    DPMatrix_ & operator=(DPMatrix_ const & other)
+//    {
+//        if (this != &other)
+//        {
+//            data_host = other.data_host;
+//        }
+//        return *this;
+//    }
```
Comment (rrahn) [inline include/seqan/align/dp_matrix.h]: Remove commented block.

File: include/seqan/find/find_myers_ukkonen.h [hunk 5]
```diff
@@ -615,109 +674,100 @@ _patternMatchNOfFinder(Pattern<TNeedle, Myers<TSpec, THasState, void> > & patter
...
+//
+//template <typename TNeedle, typename TSpec, typename THasState, typename TFindBeginPatternSpec>
+//inline TNeedle
+//host(Pattern<TNeedle, Myers<TSpec, THasState, TFindBeginPatternSpec> > & pattern)
+//{
+//SEQAN_CHECKPOINT
+//    typedef Pattern<TNeedle, Myers<TSpec, THasState, TFindBeginPatternSpec> > TPattern;
+//    return host(const_cast<TPattern const &>(pattern));
+//}
```
Comment (rrahn) [inline include/seqan/find/find_myers_ukkonen.h]: remove.

File: include/seqan/find/find_pattern_base.h [hunk 6]
```diff
@@ -221,36 +228,99 @@ struct ScoringScheme<TNeedle const>:
...
+        TNeedle2 && ndl)
+{
+    SEQAN_ASSERT(!empty(ndl));
+    setValue(_dataHost(me), std::forward<TNeedle2>(ndl));  // Set generic setHost function.
+    _reinitPattern(me);
+}
+
+template <typename TNeedle, typename TSpec, typename TNeedle2, typename TError>
+inline void
+setHost(Pattern<TNeedle, TSpec> & me,
```
Comment (rrahn) [inline include/seqan/find/find_pattern_base.h]: remove this interface as discussed above.

File: include/seqan/find/find_pattern_base.h [hunk 7]
```diff
@@ -221,36 +228,99 @@ struct ScoringScheme<TNeedle const>:
...
 {
-     me.data_host = ndl; //assign => Pattern haelt eine Kopie => doof!
+    SEQAN_ASSERT(!empty(ndl));
+    setValue(_dataHost(me), ndl);  // Set generic setHost function.
+    _reinitPattern(me);
+}
+
+template <typename TNeedle, typename TSpec, typename TNeedle2, typename TError>
+inline void
+setHost(Pattern<TNeedle, TSpec> & me,
```
Comment (rrahn) [inline include/seqan/find/find_pattern_base.h]: remove this interfaces with the distance parameter.

File: include/seqan/find/find_shiftand.h [hunk 8]
```diff
@@ -138,63 +150,64 @@ setHost(Pattern<TNeedle, ShiftAnd> & me, TNeedle2 const & needle)
...
+//    return host(pattern);
+//}
+//
+//template <typename TNeedle>
+//inline TNeedle
+//needle(Pattern<TNeedle, ShiftAnd> & pattern)
+//{
+//SEQAN_CHECKPOINT
+//    return host(const_cast<Pattern<TNeedle, ShiftAnd> const &>(pattern));
+//}
```
Comment (rrahn) [inline include/seqan/find/find_shiftand.h]: remove commented block

File: include/seqan/find/find_shiftor.h [hunk 9]
```diff
@@ -94,15 +105,17 @@ class Pattern<TNeedle, ShiftOr> {
 // Functions
 //////////////////////////////////////////////////////////////////////////////
 
-template <typename TNeedle, typename TNeedle2>
+// TODO(rrahn): Change to rvalue reference for TNeedle2 as soon as all pattern store pointer to the needle.
```
Comment (rrahn) [inline include/seqan/find/find_shiftor.h]: remove comment.

File: include/seqan/find/find_shiftor.h [hunk 10]
```diff
@@ -138,63 +151,63 @@ setHost(Pattern<TNeedle, ShiftOr> & me, TNeedle2 const & needle)
...
+//    return host(pattern);
+//}
+//
+//template <typename TNeedle>
+//inline TNeedle
+//needle(Pattern<TNeedle, ShiftOr> & pattern)
+//{
+//SEQAN_CHECKPOINT
+//    return host(const_cast<Pattern<TNeedle, ShiftOr> const &>(pattern));
+//}
```
Comment (rrahn) [inline include/seqan/find/find_shiftor.h]: remove commented block.

File: include/seqan/index/find_backtracking.h [hunk 11]
```diff
@@ -675,6 +680,7 @@ void _moveIteratorAtRoot(Pattern<Index<TNeedle, TSpec>, Backtracking<TDistance,
     me.index_iterator = TIndexIterator(host(me));
 }
 
+// TODO(rrahn): I am not sure about the index version. Should it always be copied? Seems very error prone to me!
```
Comment (rrahn) [inline include/seqan/index/find_backtracking.h]: Adapt to also allow setting a reference instead of always copying.

**Comment by esiragusa** [comment]
What is broken with non-const/const ref setHost()? What do you fix with rvalue ref setHost()?

**Comment by rrahn** [comment]
When calling setHost, it always does a const-cast on the needle, causing the underlying holder always to take a copy of the needle instead of the storing a pointer to the resource (unless the Holder is over a const TNeedle type, which I haven't seen so far.) 
If this is the intended behaviour of the pattern class then it makes no sense to use a holder at all. With universal reference, it either takes a lvalue reference and propagates it as such or it can get a temporary and propagates as such, so that by passing a rvalue the holder copies the resource and it stores only a pointer to the lvalue reference.

**Comment by esiragusa** [comment]
Thanks for clarifying! This is now fixed with C++11 only, right? Apart from Pattern, is this bug affecting any class using Holder + setHost()? To solve it properly, shouldn't we replace Holders with pointers?

**Comment by rrahn** [comment]
Right, this only fixes this issue for c++11.
There where only a few other classes that uses a holder as well:
  * Matrix
  * DPMatrix
  * JournaledString
  * JournaledSet
  * BamTags something
But for them it is just a cosmetic change in order to avoid overloading just one instead of two interfaces or simply use the default setHost.
I rather tend to explicitly manage resources outside of the Pattern class and only use weak pointer or raw pointer to the resource. But I think this involves quite some changes. We planned to refactor find module anyway. I think we can then drop the Holder member?

**Comment by esiragusa** [comment]
Ah okay, the problem arises only when one overloads setHost() and does a const cast, right?
For the find modules, it's fine like that, It has to be refactored!

**Comment by rrahn** [comment]
Yes, only the Pattern implementation did the const_cast of the non-const needle for what ever reason?! :fearful:

**Comment by esiragusa** [comment]
OK! Ship it if CI builds are fine :+1:

**Comment by rrahn** [comment]
@esiragusa I also changed some of the pattern implementations, i.e. all patterns now support a Holder.
This wasn't the case for Myers<> unbanded, ShiftAnd and ShiftOr. They encoded the needle within the bit masks and did not hold a copy of the needle. Of course it wasn't necessary, because the needle was always copied. But I don't know where and how Myers is used and if it could affect the performance of some tools with that change. For the future if we completely move back from holders to raw pointers this might not a problem anymore. What do you think?

**Comment by esiragusa** [comment]
Yep, in the new find module we will use pointers instead of holders. But it's not worth to fiddle with the old module...

**Comment by rrahn** [comment]
ok, than I will role back the changes to these 3 instances to cover the old behaviour and we change it then with the new find module.

**Comment by esiragusa** [comment]
Well if there are no issues with your changes you can keep them... :+1:

**Comment by rrahn** [comment]
Well I think all tests run through! :feelsgood:

**Pull request merged**